File: scripts/analyze-bundle.py
#!/usr/bin/env python3
"""
번들 분석 리포트에서 큰 모듈 식별
Usage: python scripts/analyze-bundle.py
"""

import json
import sys

try:
    with open('docs/bundle-data.json', 'r', encoding='utf-8') as f:
        data = json.load(f)
except FileNotFoundError:
    print("Error: docs/bundle-data.json not found")
    sys.exit(1)

modules = []

def traverse(node):
    if 'uid' in node:
        uid = node['uid']
        if uid in data['nodeParts']:
            part = data['nodeParts'][uid]
            meta_uid = part.get('metaUid')
            if meta_uid and meta_uid in data['nodeMetas']:
                meta = data['nodeMetas'][meta_uid]
                if 'id' in meta:
                    modules.append({
                        'name': meta['id'],
                        'rendered': part['renderedLength'],
                        'gzip': part['gzipLength'],
                        'brotli': part['brotliLength'],
                    })
    if 'children' in node:
        for child in node['children']:
            traverse(child)

traverse(data['tree'])

# Sort by rendered size
modules.sort(key=lambda m: m['rendered'], reverse=True)

print("=" * 80)
print("Top 10 Largest Modules (by rendered size)")
print("=" * 80)
print()

for i, m in enumerate(modules[:10], 1):
    name = m['name'].replace('C:/git/xcom-enhanced-gallery/', '')
    print(f"{i}. {name}")
    print(f"   Rendered: {m['rendered']:,} bytes ({m['rendered']/1024:.2f} KB)")
    print(f"   Gzip: {m['gzip']:,} bytes ({m['gzip']/1024:.2f} KB)")
    print(f"   Brotli: {m['brotli']:,} bytes ({m['brotli']/1024:.2f} KB)")
    print()

total_rendered = sum(m['rendered'] for m in modules)
total_gzip = sum(m['gzip'] for m in modules)
total_brotli = sum(m['brotli'] for m in modules)

top10_rendered = sum(m['rendered'] for m in modules[:10])
top10_gzip = sum(m['gzip'] for m in modules[:10])

print("=" * 80)
print("Summary")
print("=" * 80)
print()
print(f"Total modules: {len(modules)}")
print(f"Total rendered: {total_rendered:,} bytes ({total_rendered/1024:.2f} KB)")
print(f"Total gzip: {total_gzip:,} bytes ({total_gzip/1024:.2f} KB)")
print(f"Total brotli: {total_brotli:,} bytes ({total_brotli/1024:.2f} KB)")
print()
print(f"Top 10 rendered: {top10_rendered:,} bytes ({top10_rendered/1024:.2f} KB)")
print(f"Top 10 represent: {(top10_rendered/total_rendered)*100:.2f}% of total")
print()
print(f"Top 10 gzip: {top10_gzip:,} bytes ({top10_gzip/1024:.2f} KB)")
print(f"Top 10 represent: {(top10_gzip/total_gzip)*100:.2f}% of total")

# 큰 모듈 카테고리 분석 (>10 KB rendered)
print()
print("=" * 80)
print("Large Modules (>10 KB rendered)")
print("=" * 80)
print()

large_modules = [m for m in modules if m['rendered'] > 10 * 1024]
categories = {}

for m in large_modules:
    name = m['name'].replace('C:/git/xcom-enhanced-gallery/', '')
    if 'node_modules/solid-js' in name:
        category = 'Solid.js'
    elif 'src/shared/services' in name:
        category = 'Services'
    elif 'src/shared/utils' in name:
        category = 'Utils'
    elif 'src/shared/components' in name:
        category = 'Components'
    elif 'src/features' in name:
        category = 'Features'
    else:
        category = 'Other'

    if category not in categories:
        categories[category] = {'count': 0, 'rendered': 0, 'gzip': 0, 'brotli': 0}

    categories[category]['count'] += 1
    categories[category]['rendered'] += m['rendered']
    categories[category]['gzip'] += m['gzip']
    categories[category]['brotli'] += m['brotli']

for category, stats in sorted(categories.items(), key=lambda x: x[1]['rendered'], reverse=True):
    print(f"{category}:")
    print(f"  Modules: {stats['count']}")
    print(f"  Rendered: {stats['rendered']/1024:.2f} KB")
    print(f"  Gzip: {stats['gzip']/1024:.2f} KB")
    print(f"  % of total: {(stats['rendered']/total_rendered)*100:.2f}%")
    print()

# JSON 출력
output = {
    'top10': [
        {
            'name': m['name'].replace('C:/git/xcom-enhanced-gallery/', ''),
            'rendered': m['rendered'],
            'gzip': m['gzip'],
            'brotli': m['brotli']
        }
        for m in modules[:10]
    ],
    'summary': {
        'totalModules': len(modules),
        'totalRendered': total_rendered,
        'totalGzip': total_gzip,
        'totalBrotli': total_brotli,
        'top10Percent': f"{(top10_rendered/total_rendered)*100:.2f}"
    },
    'largeModules': len(large_modules),
    'categories': categories
}

with open('docs/bundle-analysis-summary.json', 'w', encoding='utf-8') as f:
    json.dump(output, f, indent=2)

print("Summary saved to: docs/bundle-analysis-summary.json")

File: scripts/analyze-bundle.js
#!/usr/bin/env node

/**
 * @fileoverview Phase 326.5: Bundle Analysis Script
 * @description Analyze production bundle for optimization opportunities
 *
 * Usage: node scripts/analyze-bundle.js
 * Output: Terminal report + docs/phase-326-bundle-analysis.json
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const projectRoot = path.resolve(__dirname, '..');
const distDir = path.resolve(projectRoot, 'dist');
const docDir = path.resolve(projectRoot, 'docs');

/**
 * Get file size in human-readable format
 */
function formatBytes(bytes) {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + ' ' + sizes[i];
}

/**
 * Calculate gzip compression size (estimation)
 */
function estimateGzipSize(bytes) {
  // Rough estimation: minified JS typically compresses to 30-40% of original
  // Minified + gzip typically achieves 20-30% compression ratio
  return Math.round(bytes * 0.25); // Conservative 25% estimate
}

/**
 * Analyze production bundle
 */
async function analyzeBundle() {
  console.log('\n📊 Phase 326.5: Bundle Analysis\n');
  console.log('='.repeat(60));

  try {
    // Check if bundle exists
    const bundlePath = path.resolve(distDir, 'xcom-enhanced-gallery.user.js');
    if (!fs.existsSync(bundlePath)) {
      console.error('❌ Bundle not found. Run: npm run build:only');
      process.exit(1);
    }

    // Read bundle
    const bundleContent = fs.readFileSync(bundlePath);
    const bundleSize = bundleContent.length;
    const bundleSizeHuman = formatBytes(bundleSize);

    // Calculate gzip estimate
    const gzipEstimate = estimateGzipSize(bundleSize);
    const gzipHuman = formatBytes(gzipEstimate);

    // Read CSS file if exists
    const cssFiles = fs
      .readdirSync(distDir)
      .filter(f => f.startsWith('assets/style-') && f.endsWith('.css'));
    let cssSize = 0;
    let cssHuman = '0 B';
    if (cssFiles.length > 0) {
      const cssPath = path.resolve(distDir, cssFiles[0]);
      cssSize = fs.statSync(cssPath).size;
      cssHuman = formatBytes(cssSize);
    }

    // Read JS asset file if exists
    const jsFiles = fs
      .readdirSync(distDir)
      .filter(f => f.startsWith('assets/main-') && f.endsWith('.js'));
    let jsAssetSize = 0;
    let jsAssetHuman = '0 B';
    if (jsFiles.length > 0) {
      const jsPath = path.resolve(distDir, jsFiles[0]);
      jsAssetSize = fs.statSync(jsPath).size;
      jsAssetHuman = formatBytes(jsAssetSize);
    }

    // Calculate line count (rough estimate)
    const bundleLines = bundleContent.toString().split('\n').length;

    // Generate report
    const report = {
      timestamp: new Date().toISOString(),
      phase: '326.5-1',
      baseline: {
        bundleSize: {
          bytes: bundleSize,
          human: bundleSizeHuman,
        },
        gzipEstimate: {
          bytes: gzipEstimate,
          human: gzipHuman,
          ratio: `${((gzipEstimate / bundleSize) * 100).toFixed(1)}%`,
        },
        cssAsset: {
          bytes: cssSize,
          human: cssHuman,
        },
        jsAsset: {
          bytes: jsAssetSize,
          human: jsAssetHuman,
        },
        lineCount: bundleLines,
      },
      features: {
        gallery: {
          estimatedSize: '120 KB',
          required: true,
          optimization: 'Optimize media extraction',
        },
        settings: {
          estimatedSize: '80 KB',
          required: true,
          optimization: 'Conditional panel loading',
        },
        download: {
          estimatedSize: '60 KB',
          required: true,
          optimization: 'ZIP/batch optimization',
        },
        mediaExtraction: {
          estimatedSize: '50 KB',
          required: false,
          optimization: 'Tree-shake if disabled',
        },
        advancedFilters: {
          estimatedSize: '40 KB',
          required: false,
          optimization: 'Conditional UI loading',
        },
        accessibility: {
          estimatedSize: '35 KB',
          required: false,
          optimization: 'WCAG compliance module',
        },
        sharedUtils: {
          estimatedSize: '22 KB',
          required: true,
          optimization: 'Deduplication analysis',
        },
      },
      target: {
        currentSize: bundleSize,
        targetSize: 320512, // 310-315 KB target (using 312.5 KB = 320000 bytes)
        targetHuman: '310-315 KB',
        potentialReduction: {
          bytes: bundleSize - 320512,
          percentage: (((bundleSize - 320512) / bundleSize) * 100).toFixed(1),
        },
      },
      optimizationStrategy: [
        'Feature-based tree-shaking (optional features)',
        'Shared utility consolidation',
        'CSS unused class purging',
        'Dynamic import optimization',
      ],
      nextPhase: '326.5-2 (Bundle Analysis Details)',
    };

    // Display report
    console.log('\n📈 BUNDLE BASELINE REPORT\n');

    console.log('Core Metrics:');
    console.log(
      `  Production Bundle:  ${report.baseline.bundleSize.human} (${report.baseline.bundleSize.bytes.toLocaleString()} bytes)`
    );
    console.log(
      `  Gzip Estimate:      ${report.baseline.gzipEstimate.human} (${report.baseline.gzipEstimate.ratio})`
    );
    console.log(`  CSS Assets:         ${report.baseline.cssAsset.human}`);
    console.log(`  JS Assets:          ${report.baseline.jsAsset.human}`);
    console.log(`  Line Count:         ${report.baseline.lineCount.toLocaleString()} lines`);

    console.log('\nOptimization Target:');
    console.log(`  Target Size:        ${report.target.targetHuman}`);
    console.log(
      `  Reduction Needed:   ${formatBytes(report.target.potentialReduction.bytes)} (${report.target.potentialReduction.percentage}%)`
    );

    console.log('\nOptional Features (tree-shaking candidates):');
    console.log('  - mediaExtraction (50 KB)');
    console.log('  - advancedFilters (40 KB)');
    console.log('  - accessibility (35 KB)');
    console.log('  Total Optional: 125 KB');

    console.log('\nOptimization Strategy:');
    report.optimizationStrategy.forEach(strategy => {
      console.log(`  ✓ ${strategy}`);
    });

    console.log(`\n✅ Phase 326.5-1: Baseline Complete`);
    console.log(`📊 Next: Phase 326.5-2 (Detailed Bundle Analysis)\n`);

    // Save report
    const reportPath = path.resolve(docDir, 'phase-326-bundle-analysis.json');
    fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));
    console.log(`📄 Report saved: ${reportPath}\n`);

    console.log('='.repeat(60) + '\n');

    return report;
  } catch (error) {
    console.error('❌ Error during bundle analysis:', error.message);
    process.exit(1);
  }
}

// Run analysis
await analyzeBundle();
